### PR TITLE
Execute provided script before every test (`--bootstrap`)

### PR DESCRIPTION
**Pull request by jkuchar**
- bug fix? no
- new feature? yes
- BC break? no
- doc PR: 

There should be some global way how to configure environment of tests. Yep I'm looking at you `Environment::setup()`. 

- `--bootstrap` script should configure only environment only in way that tests can run without it
- `Environment::setup()` is probably good candidate for bootstrap script (at least coverage setup part)
- In future this can be used to simplify running coverage / setting up custom coverage collector (which needs to be done inside test but has no value for test itself)
- In contract putting auto-loader there is **very bad idea** as tests are not standalone anymore
- I have added `stdout` to `PhpInterpreter` exception when execution error occurs; this simplifies debugging if bootstrap fails for some reason

Discussion: is `--bootstrap` sufficient name? In real it should not bootstrap test, it should bootstrap test environment for Tester. It should have (almost) no connection to test itself.

### CODE REVIEW

File: src/Runner/CliTester.php [hunk 1]
```diff
@@ -54,6 +54,11 @@ public function run()
 			return;
 		}
 
+		if (isset($this->options['--bootstrap'])) {
+			$exportedPath = var_export($this->options['--bootstrap'], TRUE);
+			$this->interpreter->addPhpIniOption('auto_prepend_file', $exportedPath);
```
Comment (jkuchar) [inline src/Runner/CliTester.php]: This should work also with HHVM [test](https://github.com/facebook/hhvm/blob/46a9a688340dd4f38806c4ca1e5798c13fb5264c/hphp/test/slow/program_functions/auto_prepend_file-1.php.ini)

**Comment by dg** [comment]
You can use `-d auto_prepend_file=file`

**Comment by jkuchar** [comment]
You are right, that is how it works anyway.

So how do I get to this idea...

I want to make coverage collector more versatile, e.g. to support branch coverage or performace data collection, this needs to be set-up somewhere. And only place this can be set up is inside test's environment.

You can say, lets put it into bootstrap.php, which exists in every test suite by convention. And you are right. Technically you can do it. But this code has nothing to do with test logic itself, it is Tester internal infrastructure code? So I want to provide tool which seems legit (=has intuitive name), where you configure these things and do not mess your tests `bootstrap.php` with this tester-related infrastructure code.

Btw when you then want to switch coverage generator which depends on custom coverage collector (e.g. branch coverage collector), you should be able to do that just by providing configuration options to Tester CLI. Does that make sense?

One more idea - you can write completely Tester independent tests in Tester (e.g.`<?php if(doSth() === 5) exit(1);`). This test has no dependencies, and it rund inside Tester without any additional code. Until... Until you try to run coverage. Then Tester needs its infrastructure code to be in place for every test from test suite, otherwise coverage report will not include tests without coverage collector. (BTW there is currently no way how to detect these tests) Important to mention is that injected code must not affect result or execution of test, it just prepares PHP's environment to collect coverage, that's it.

I was considering to make `--bootstrap=autostart-coverage.php` as default value, which will just make sure that every test will have coverage code injected.

TL;DR: I want to make this configurable because I want to have easy to use injection point for configuring custom coverage collector.

**Comment by jkuchar** [comment]
Maybe rename to `--coverage-collector=file`? This file will be included only when coverage is enabled in Tester and there can be default implementation which inits default coverage collector. This will allow Tester to run coverage on any test (even completely independent ones) and in future configure custom coverage collector.

Downside is that there is no way how `--coverage-collector` can configure environment to pretty print error messages as it is run ony when coverage is run. (e.g. to pretty print exception from tests that are completely Tester independent)

**Comment by dg** [comment]
Related to #275

**Comment by milo** [comment]
@jkuchar It is a long reading, I hope I understand.

You want to introduce some kind of setup tasks, so test will run like:
```php
require 'setup-task-1.php';
require 'setup-task-2.php';
require 'test-itself.phpt';
```

Besides the concept, there is a one technical problem. You can have only one `auto_prepend_file`. So multiple tasks must be multiplexed into single file.

According to `--bootstrap` option, I would prefer to use current `--setup`. Something like:
```php
$runner->addSetupScript('path/to/task-1.php');
$runner->addSetupScript('path/to/task-2.php');
```

Next question is, how to multiplex it into single script:
1) Create script in temporary directory. It is problematic, Tester does not use any temp dir. And not sure about `open_basedir` in CLI.
2) Use some script embeded into Tester, some `prepend.php` and pass scripts via environment:
```
TESTER_SETUP_SCRIPT_COUNT=2
TESTER_SETUP_SCRIPT_1=path/to/task-1.php
TESTER_SETUP_SCRIPT_2=path/to/task-1.php
```
3) ???

In any case, `auto_prepend_file` can be only one. So any mechanism using it must take user's or system's `auto_prepend_file` setting in to mind.

Besides all of this, functionality to run test as a simple script by `php test.phpt` has to be kept.

**Comment by dg** [comment]
I'm closing PR, it's not a much requested feature.
